### PR TITLE
Management Command tool to scrub references to the username in form metadata/xml

### PR DESCRIPTION
**Pull request by pr33thi**
Task #2 in this doc:
https://docs.google.com/document/d/1qCtWmQ2disUEi86Ty5JwNS_3rFiveMDpZf0cj0zVcvQ/edit?ts=5aafed1b

- Gets rid of the username from the xml (and therefore the metadata) (SQL)
- Gets rid of the username from the form metadata (couch)
^^Replace username in both cases with "Redacted User (GDPR)"

Note: Saving the operation doesn't work/ doesn't add a form history tab, feedback on that would be great @snopoke or anyone familiar

### CODE REVIEW

File: corehq/apps/users/management/commands/gdpr_scrub_user.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        # Get the xml attachment from the form data
+        form_attachment_xml = form_data.get_attachment("form.xml")
+
+        # Convert the xml string to dict
+        form_attachment_dict = xmltodict.parse(form_attachment_xml)
+
+        # Replace the old username with the new username
+        form_attachment_dict["data"]["n0:meta"]["n0:username"] = new_username
+        # Convert the dict back to xml
+        form_attachment_xml_new = xmltodict.unparse(form_attachment_dict)
```
Comment (millerdev) [inline corehq/apps/users/management/commands/gdpr_scrub_user.py]: How similar is the new XML to the original? Are all namespaces and other XML features preserved? This seems like a rather complex and possibly lossy operation, and I'm surprised it can be done so simply given that XML does not really have a canonical form. I am not familiar with `xmltodict` at all, fwiw.

File: corehq/apps/users/management/commands/gdpr_scrub_user.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    def handle(self, user_id, domain, **options):
+        this_form_accessor = FormAccessors(domain=domain)
+        form_ids = this_form_accessor.get_form_ids_for_user(user_id)
+        new_username = "Redacted User (GDPR)"
+        for form_data in this_form_accessor.iter_forms(form_ids):
+            form_attachment_xml_new = self.parse_form_data(form_data, new_username)
+            this_form_accessor.modify_attachment_xml_and_metadata(form_data, form_attachment_xml_new)
+
+    @staticmethod
+    def parse_form_data(form_data, new_username):
```
Comment (millerdev) [inline corehq/apps/users/management/commands/gdpr_scrub_user.py]: Nitpick: what do you think of renaming this to `update_form_data` since it doesn't just parse the form?

File: corehq/form_processor/tests/test_gdpr_scrub_user.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    @use_sql_backend
+    def test_modify_attachment_xml_and_metadata_sql(self):
+        form = get_simple_wrapped_form(uuid.uuid4().hex,
+                                       metadata=TestFormMetadata(domain=DOMAIN),
+                                       simple_form=GDPR_SIMPLE_FORM)
+        new_form_xml = Command().parse_form_data(form, self.new_username)
+        FormAccessors(DOMAIN).modify_attachment_xml_and_metadata(form, new_form_xml)
+
+        # Test that the xml changed
+        form_attachment_xml = form.get_attachment("form.xml")
```
Comment (millerdev) [inline corehq/form_processor/tests/test_gdpr_scrub_user.py]: Would be nice to compare this new/updated XML to the original `simple_form` (mutated with the new username) to ensure that the new XML matches the old using something different than the library that you're making the original mutation with. We have [an assert method that does that](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/tests/util.py#L30).

**Review by millerdev — COMMENTED** [review]
Looking good. I added a few comments.

File: corehq/apps/users/management/commands/gdpr_scrub_user.py (hunk 1 above)
Comment (pr33thi) [reply corehq/apps/users/management/commands/gdpr_scrub_user.py]: Switched form xmltodict to lxml here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/c7b17ed7e4ae91458fd2fdbad435222bceeb921e

File: corehq/form_processor/tests/test_gdpr_scrub_user.py (hunk 3 above)
Comment (pr33thi) [reply corehq/form_processor/tests/test_gdpr_scrub_user.py]: Fixed here (was much easier with lxml instead of xmltodict):
https://github.com/dimagi/commcare-hq/pull/19979/commits/6e77f07d8078d49dc91fdfed0c281488ab6a508e

File: corehq/apps/users/management/commands/gdpr_scrub_user.py (hunk 2 above)
Comment (pr33thi) [reply corehq/apps/users/management/commands/gdpr_scrub_user.py]: Good idea, changed here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/8d6837e0473a99499db893b8e79ac0d7e1afb62b

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import unicode_literals
+from django.core.management.base import BaseCommand
+from corehq.form_processor.interfaces.dbaccessors import FormAccessors
+from StringIO import StringIO
```
[the remark this answers is not in the document]
Comment (pr33thi) [reply corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: @nickpell 
what does this mean?

[the remark this answers is not in the document]
Comment (nickpell) [reply corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: You'll want to replace ```from StringIO import StringIO``` with ```from io import StringIO``` (or ```BytesIO``` if you're dealing with byte strings).

Python 3 uses the ```io``` module, and removes the ```StringIO``` module.

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    def add_arguments(self, parser):
+        parser.add_argument('username')
+        parser.add_argument('domain')
+
+    def handle(self, username, domain, **options):
+        this_form_accessor = FormAccessors(domain=domain)
+        user = CommCareUser.get_by_username(username)
+        if not user:
+            logger.info("User {} not found.".format(username))
+            sys.exit()
```
Comment (millerdev) [inline corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: should say `sys.exit(1)`, which indicates an error condition

File: corehq/form_processor/backends/couch/dbaccessors.py [hunk 6]
```diff
@@ -126,6 +131,14 @@ def _form_undelete(doc):
             doc['server_modified_on'] = json_format_datetime(datetime.utcnow())
         return _soft_undelete(XFormInstance.get_db(), form_ids, _form_undelete)
 
+    @staticmethod
+    def modify_attachment_xml_and_metadata(form_data, form_attachment_new_xml):
+        form_data.put_attachment(form_attachment_new_xml, name="form.xml", content_type='text/xml')
+        operation = XFormOperation(user_id=SYSTEM_USER_ID, date=datetime.utcnow(),
+                                   operation='Scrub username for GDPR compliance.')
+        form_data.history.append(operation)  # TODO: should this show in Form History tab? it doesn't
```
Comment (millerdev) [inline corehq/form_processor/backends/couch/dbaccessors.py]: I think this TODO should be resolved before the PR is merged if possible. What do you think?

File: corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        # Test that the xml changed
+        actual_form_xml = form.get_attachment("form.xml")
+        self.assertXMLEqual(EXPECTED_FORM_XML, actual_form_xml)
+
+        # Test that the metadata changed in the database
+        attachment_metadata = form.get_attachment_meta("form.xml")
+        actual_form_xml_from_db = XFormAttachmentSQL.read_content(attachment_metadata)
+        self.assertXMLEqual(EXPECTED_FORM_XML, actual_form_xml_from_db)
+
+        # TODO: Test that the operations history is updated
```
Comment (millerdev) [inline corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py]: Are you planning to do this before the PR is merged?

**Review by millerdev — COMMENTED** [review]
One nitpicky comment and two questions. Looking good.

File: corehq/form_processor/backends/couch/dbaccessors.py (hunk 6 above)
Comment (proteusvacuum) [reply corehq/form_processor/backends/couch/dbaccessors.py]: check out this comment to see how you might achieve this: https://github.com/dimagi/commcare-hq/pull/19973#discussion_r177989201

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        user_id = user._id
+        form_ids = this_form_accessor.get_form_ids_for_user(user_id)
+        new_username = "Redacted User (GDPR)"
+        forms_updated = 0
+        for form_data in this_form_accessor.iter_forms(form_ids):
+            form_attachment_xml_new = self.update_form_data(form_data, new_username)
+            this_form_accessor.modify_attachment_xml_and_metadata(form_data, form_attachment_xml_new)
+            forms_updated += 1
+        logging.info("Updated {} forms for user {}".format(forms_updated, username))
```
Comment (esoergel) [inline corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: What do you think about putting in a confirmation here?  Something like "Update 132 forms on 'domain' for username@domain.commcarehq.org? (Y/n)".  These commands are intentionally destructive and misuses would be unrecoverable, so we should be pretty cautious.

File: corehq/form_processor/backends/couch/dbaccessors.py [hunk 9]
```diff
@@ -102,6 +103,10 @@ def get_attachment_content(form_id, attachment_id):
     def save_new_form(form):
         form.save()
 
+    @staticmethod
+    def update_form(form):
+        form.save()
```
Comment (esoergel) [inline corehq/form_processor/backends/couch/dbaccessors.py]: Is this (and the superclass version) used anywhere?

File: corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    <n0:meta xmlns:n0="http://openrosa.org/jr/xforms">
+        <n0:deviceID>{device_id}</n0:deviceID>
+        <n0:timeStart>{time_start}</n0:timeStart>
+        <n0:timeEnd>{time_end}</n0:timeEnd>
+        <n0:username>{username}</n0:username>
+        <n0:userID>{user_id}</n0:userID>
+        <n1:appVersion xmlns:n1="http://commcarehq.org/xforms"></n1:appVersion>
+    </n0:meta>
+    {case_block}
+</data>"""
```
Comment (esoergel) [inline corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py]: Any reason not to just use the existing `SIMPLE_FORM` template?  Looks like the differences are `n0` instead of `n1`, and the `instanceID`, which both seems tangential to this test.

**Review by esoergel — APPROVED** [review]
Is there going to be a second command to remove the other references to `username` on the form db object?

File: corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py (hunk 7 above)
Comment (pr33thi) [reply corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py]: Yep! I mentioned I'm going to finish this in the PR description (but that's way up there now)

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    @staticmethod
+    def update_form_data(form_data, new_username):
```
Comment (snopoke) [inline corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: a string search and replace might be faster

File: corehq/form_processor/backends/couch/dbaccessors.py (hunk 6 above)
Comment (snopoke) [reply corehq/form_processor/backends/couch/dbaccessors.py]: This is operating on the couch models so this approach should be fine. A unit test should be easy to make sure it's working

File: corehq/form_processor/backends/sql/dbaccessors.py [hunk 12]
```diff
@@ -581,6 +582,17 @@ def soft_undelete_forms(domain, form_ids):
 
         return return_value
 
+    @staticmethod
+    def modify_attachment_xml_and_metadata(form_data, form_attachment_new_xml):
+        attachment_metadata = form_data.get_attachment_meta("form.xml")
+        # Write the new xml to the database
+        XFormAttachmentSQL.write_content(attachment_metadata, form_attachment_new_xml)
```
Comment (snopoke) [inline corehq/form_processor/backends/sql/dbaccessors.py]: why not `attachment_metadata.write_content(form_attachment_new_xml)`

File: corehq/form_processor/backends/sql/dbaccessors.py [hunk 13]
```diff
@@ -581,6 +582,17 @@ def soft_undelete_forms(domain, form_ids):
...
+        attachment_metadata.save()
+        operation = XFormOperationSQL(user_id=SYSTEM_USER_ID, date=datetime.utcnow(),
+                                      operation='Scrub username for GDPR compliance.')
+        form_data.history.append(operation)  # TODO: should this show in Form History tab? it doesn't
```
Comment (snopoke) [inline corehq/form_processor/backends/sql/dbaccessors.py]: this won't work for SQL. Rather do this:
```python
form_data.track_create(operation)
FormAccessorSQL.update_form(form_data)
```

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py (hunk 11 above)
Comment (pr33thi) [reply corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: True, I was imagining a dev would just run this in the background while doing other things anyway (as it'll probably end up taking a while either way) and searching the xml seems a little more robust.

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py (hunk 5 above)
Comment (pr33thi) [reply corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: https://github.com/dimagi/commcare-hq/pull/19979/commits/4b45c9117c40d43ca06f3d5dfeaae87f0d68cc38

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py (hunk 8 above)
Comment (pr33thi) [reply corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: Good call, added it in here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/dd1468d1e5d288001f6157d643795b415839b499

File: corehq/form_processor/backends/couch/dbaccessors.py (hunk 9 above)
Comment (pr33thi) [reply corehq/form_processor/backends/couch/dbaccessors.py]: Nope! This is a remnant from something I tried earlier that isn't relevant anymore. 
Removed them both here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/3785fe4ab63bca7fa19dd033723a144151077093

File: corehq/form_processor/backends/sql/dbaccessors.py (hunk 12 above)
Comment (pr33thi) [reply corehq/form_processor/backends/sql/dbaccessors.py]: True, fixed here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/d42272370e58270d60fc2b82815e0ad48fdab195

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    def handle(self, username, domain, **options):
+        this_form_accessor = FormAccessors(domain=domain)
+        user = CommCareUser.get_by_username(username)
+        if not user:
+            logger.info("User {} not found.".format(username))
+            sys.exit(1)
+        user_id = user._id
+        form_ids = this_form_accessor.get_form_ids_for_user(user_id)
+        new_username = "Redacted User (GDPR)"
+        input_response = raw_input("Update {} form(s) for user {} in domain {}? (y/n): ".format(len(form_ids), username, domain))
```
[the remark this answers is not in the document]
Comment (nickpell) [reply corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: This is fixed with ```six```: https://pythonhosted.org/six/#module-six.moves

File: corehq/form_processor/backends/couch/dbaccessors.py (hunk 6 above)
Comment (pr33thi) [reply corehq/form_processor/backends/couch/dbaccessors.py]: Tested here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/8f4cf33e4b0319561c0110c5bb470e9790b3bb27

File: corehq/form_processor/backends/sql/dbaccessors.py (hunk 13 above)
Comment (pr33thi) [reply corehq/form_processor/backends/sql/dbaccessors.py]: Updated here (and did some extra test stuff which I clean up in the below pr):
https://github.com/dimagi/commcare-hq/pull/19979/commits/c38433c08998c958cf9e44cdd6f5a26248314aa8

And tested here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/8f4cf33e4b0319561c0110c5bb470e9790b3bb27

File: corehq/form_processor/models.py [hunk 15]
```diff
@@ -343,7 +343,7 @@ def form_data(self):
     def history(self):
         """:returns: List of XFormOperationSQL objects"""
         from corehq.form_processor.backends.sql.dbaccessors import FormAccessorSQL
-        operations = FormAccessorSQL.get_form_operations(self.form_id) if self.is_saved() else []
+        operations = FormAccessorSQL.get_form_operations(self.form_id) if self.is_saved() else ['hi!!!!']
```
Comment (snopoke) [inline corehq/form_processor/models.py]: `hi!!!!`

File: corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+class GDPRScrubUserFromFormsCouchTests(TestCase):
+    def setUp(self):
+        super(GDPRScrubUserFromFormsCouchTests, self).setUp()
+        self.db = TemporaryFilesystemBlobDB()
+
+    def tearDown(self):
+        self.db.close()
+        super(GDPRScrubUserFromFormsCouchTests, self).tearDown()
+
+    def test_modify_attachment_xml_and_metadata_couch(self):
```
Comment (snopoke) [inline corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py]: These tests look the same. The common pattern is to do this:

```
class CouchTests(TestCase):
    def my_test(self):
        ....

@use_sql_backend
class SQLtests(CouchTests)
    pass


```

**Review by snopoke — CHANGES_REQUESTED** [review]

**Review by millerdev — COMMENTED** [review]
👍 after @snopoke's feedback is addressed

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (pr33thi) [reply corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: Thanks! Fixed here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/77a4504e821f0beef78f42976bc50fb547df0afb

File: corehq/form_processor/backends/sql/dbaccessors.py [hunk 17]
```diff
@@ -581,6 +582,17 @@ def soft_undelete_forms(domain, form_ids):
...
         return return_value
 
+    @staticmethod
+    def modify_attachment_xml_and_metadata(form_data, form_attachment_new_xml):
+        attachment_metadata = form_data.get_attachment_meta("form.xml")
+        # Write the new xml to the database
+        attachment_metadata.write_content(form_attachment_new_xml)
+        attachment_metadata.save()
+        operation = XFormOperationSQL(user_id=SYSTEM_USER_ID, date=datetime.utcnow(),
+                                      operation='Scrub username for GDPR compliance.')
```
[the remark this answers is not in the document]
Comment (pr33thi) [reply corehq/form_processor/backends/sql/dbaccessors.py]: Ah, didnt know about that, using that logic here:

https://github.com/dimagi/commcare-hq/pull/19979/commits/8da6f23e2660f1d6ed3e0c1cd86a6965055c3f43

And just to be clear, `XFormOperation` doesn't have those terms predefined, right? I just passed in `'gdpr_scrub'` directly here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/34a94c2a68078a5d77e1bbbaad68465d5120f2d9#diff-ff25fb2af7d300989db9379ec9dd37fcL134

File: corehq/form_processor/models.py (hunk 15 above)
Comment (pr33thi) [reply corehq/form_processor/models.py]: oops, fixed here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/94ca91c82202a27348915facb2c53e01b4c1c263

File: corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py (hunk 16 above)
Comment (pr33thi) [reply corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py]: cool, didn't know I could do that, fixed here:

https://github.com/dimagi/commcare-hq/pull/19979/commits/34a94c2a68078a5d77e1bbbaad68465d5120f2d9

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+class Command(BaseCommand):
+    help = "Scrubs the username from all forms associated with the given user"
+
+    def add_arguments(self, parser):
+        parser.add_argument('username')
+        parser.add_argument('domain')
+
+    def handle(self, username, domain, **options):
+        this_form_accessor = FormAccessors(domain=domain)
+        user = CommCareUser.get_by_username(username)
```
Comment (snopoke) [inline corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: This could be a web user too. Probably better to use `CouchUser.get_by_username`

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+        if not user:
+            logger.info("User {} not found.".format(username))
+            sys.exit(1)
+        user_id = user._id
+        form_ids = this_form_accessor.get_form_ids_for_user(user_id)
+        new_username = "Redacted User (GDPR)"
```
Comment (snopoke) [inline corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: might be nice to have this as a constant

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+        else:
+            logging.info("Command not recognized. Exiting.")
+
+    @staticmethod
+    def update_form_data(form_data, new_username):
+        form_attachment_xml = form_data.get_attachment("form.xml")
+        xml_elem = etree.parse(StringIO(six.text_type(form_attachment_xml)))
+        id_elem = xml_elem.find("{http://openrosa.org/jr/xforms}meta").find(
+            "{http://openrosa.org/jr/xforms}username")
+        id_elem.text = new_username
```
Comment (snopoke) [inline corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: I still think string search and replace would be faster and more robust here. It is possible that a username could be saved to any question in a form.

File: corehq/form_processor/backends/couch/dbaccessors.py [hunk 21]
```diff
@@ -126,6 +127,14 @@ def _form_undelete(doc):
             doc['server_modified_on'] = json_format_datetime(datetime.utcnow())
         return _soft_undelete(XFormInstance.get_db(), form_ids, _form_undelete)
 
+    @staticmethod
+    def modify_attachment_xml_and_metadata(form_data, form_attachment_new_xml):
+        form_data.put_attachment(form_attachment_new_xml, name="form.xml", content_type='text/xml')
+        operation = XFormOperation(user_id=SYSTEM_USER_ID, date=datetime.utcnow(),
+                                   operation='gdpr_scrub')
+        form_data.history.append(operation)  # TODO: should this show in Form History tab? it doesn't
```
Comment (snopoke) [inline corehq/form_processor/backends/couch/dbaccessors.py]: is this comment still true?

File: corehq/form_processor/backends/couch/dbaccessors.py [hunk 22]
```diff
@@ -126,6 +127,14 @@ def _form_undelete(doc):
             doc['server_modified_on'] = json_format_datetime(datetime.utcnow())
         return _soft_undelete(XFormInstance.get_db(), form_ids, _form_undelete)
 
+    @staticmethod
+    def modify_attachment_xml_and_metadata(form_data, form_attachment_new_xml):
```
Comment (snopoke) [inline corehq/form_processor/backends/couch/dbaccessors.py]: this still needs to modify the form metadata for couch forms

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py (hunk 20 above)
Comment (millerdev) [reply corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: Weighing in here because I think an unstructured search/replace has some significant disadvantages. Suppose a project is scrubbing username `test@test.com` and there's another user in the domain named `my_test@test.com`—unless crafted very carefully (and email addresses are notoriously hard to parse) a search/replace may match things that should not be replaced. Also, unstructured search/replace could cause other parts of the metadata (or anything stored in couch or SQL or elastic that has been parsed from the XML) to become out of sync with the XML.

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py (hunk 18 above)
Comment (pr33thi) [reply corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: Fixed here (and learned the differences between types of users +1)
https://github.com/dimagi/commcare-hq/pull/19979/commits/de2e98ad22ae74f559608bbb2f0a2194a7599cdf

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py (hunk 19 above)
Comment (pr33thi) [reply corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: Done here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/9782b6ada66f767615c35af61d31483a9a562b70

File: corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py (hunk 20 above)
Comment (pr33thi) [reply corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py]: Agreed ^, besides I checked with @krsdimagi who confirmed that instances of the username not specifically in the `username` tag are not in our scope

File: corehq/form_processor/backends/couch/dbaccessors.py (hunk 21 above)
Comment (pr33thi) [reply corehq/form_processor/backends/couch/dbaccessors.py]: Nope, thanks for catching that
Removed here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/39339c57c1a196f76b02b81d5f29ed0e4859a9b7

File: corehq/form_processor/backends/couch/dbaccessors.py (hunk 22 above)
Comment (pr33thi) [reply corehq/form_processor/backends/couch/dbaccessors.py]: Fixed across these 2 commits:

https://github.com/dimagi/commcare-hq/pull/19979/commits/39575949e103611d51e4b0a962ffb7528dda8467

https://github.com/dimagi/commcare-hq/pull/19979/commits/856c22edd5a2e6a8def96fa90f4510f04da6754d

File: corehq/form_processor/backends/couch/dbaccessors.py [hunk 23]
```diff
@@ -126,6 +127,18 @@ def _form_undelete(doc):
...
         return _soft_undelete(XFormInstance.get_db(), form_ids, _form_undelete)
 
+    @staticmethod
+    def modify_attachment_xml_and_metadata(form_data, form_attachment_new_xml, new_username):
+        # Update XML
+        form_data.put_attachment(form_attachment_new_xml, name="form.xml", content_type='text/xml')
+        operation = XFormOperation(user_id=SYSTEM_USER_ID, date=datetime.utcnow(),
+                                   operation='gdpr_scrub')
+        form_data.history.append(operation)
+        form_data.save()
```
Comment (millerdev) [inline corehq/form_processor/backends/couch/dbaccessors.py]: nit: this save could be omitted since it's saved again just below.

**Review by millerdev — APPROVED** [review]

**Review by snopoke — APPROVED** [review]
pending resolution to double save comment: https://github.com/dimagi/commcare-hq/pull/19979/files#r179637693

**Review by millerdev — APPROVED** [review]

**Comment by millerdev** [comment]
🎉
